core/digest.go

```diff
@@ -0,0 +1,228 @@
+package core
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"reflect"
+	"sort"
+
+	"github.com/moby/buildkit/solver/pb"
+	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+)
+
+func stableDigest(value any) (digest.Digest, error) {
```

Comment:
If we stick with this approach I think we need some memoization, otherwise afaict I think we walk every possible path in the DAG which has complexity bad enough to the point that it would just freeze for a reasonable size number of vertexes+edges.

Reply:
Nevermind, tests are passing now quickly without this, doesn't appear to be needed